### PR TITLE
Commented the decompressors (and some misc things)

### PR DESCRIPTION
**Pull request by GabrielRavier**
I think I've done them all. Kosinski, Enigma and Nemesis documentation all commented in their entirity (though I've partially copied Sega Retro on that one + the Sonic 1 disasm comments)

I've also commented 2-3 other things, the controller reading for exemple.

I'll add that if I made errors (or you think my comments blocks are too big), you can actually correct them as I've allowed edits from maintainers.

### CODE REVIEW

File: s2.asm [hunk 1]
```diff
@@ -1232,8 +1232,8 @@ Joypad_Read:
 
 ; sub_1158:
 VDPSetupGame:
-	lea	(VDP_control_port).l,a0
-	lea	(VDP_data_port).l,a1
+	lea	(VDP_control_port).l,a0	; load VDP control port into a0
+	lea	(VDP_data_port).l,a1	; load VDP data port into a1
```
Comment (Clownacy) [inline s2.asm]: Seems pretty pointless.

File: s2.asm [hunk 2]
```diff
@@ -1767,54 +1787,69 @@ NemDec_WriteAndAdvance_XOR:
...
+	bra.s	Nem_BCT_Loop	; repeat
+; ---------------------------------------------------------------------------
+
+; the Nemesis decompressor uses prefix-free codes (no valid code is a prefix of a longer code)
+; e.g. if 10 is a valid 2-bit code, 110 is a valid 3-bit code but 100 isn't
+; also, when the actual compressed data is processed the high bit of each code is in bit position 7
+; so the code needs to be bit-shifted appropriately over here before being used as a code table index
+; additionally, the code needs multiple entries in the table because no masking is done during compressed data processing
+; so if 11000 is a valid code then all indices of the form 11000XXX need to have the same entry
 loc_1606:
```
Comment (Clownacy) [inline s2.asm]: Legacy labels should be commented-out.

File: s2.asm [hunk 3]
```diff
@@ -2293,88 +2335,99 @@ EniDec_ChkGetNextByte:
...
+	move.w	(sp),d5	; copy first description field
+	moveq	#$F,d4	; 16 bits in a byte
 
--
-	lsr.w	#1,d5
+Kos_Decomp_Loop:
+	lsr.w	#1,d5	; bit which is shifted out goes into C flag
 	move	sr,d6
-	dbf	d4,+
+	dbf		d4,Kos_Decomp_ChkBit
```
Comment (Clownacy) [inline s2.asm]: As s2.notes.txt says, the disassembly was set up with a tab size of 8 in mind, so this appears wrong.

File: s2.asm [hunk 4]
```diff
@@ -176,7 +176,7 @@ ErrorTrap:
 ; loc_206:
 EntryPoint:
 	tst.l	(HW_Port_1_Control-1).l	; test ports A and B control
-	bne.s	PortA_Ok	; If so, branch.
+	bne.s	PortA_Ok	; If so, magically branch.
```
Comment (LukeUsher) [inline s2.asm]: Why magically? I believe this just checks if the IO ports have been initialised already, and skips the init process if so, nothing magic about it.

File: s2.asm [hunk 5]
```diff
@@ -2293,88 +2335,99 @@ EniDec_ChkGetNextByte:
...
 ; ---------------------------------------------------------------------------
-+
+
+Kos_Decomp_Match:
 	moveq	#0,d3
-	lsr.w	#1,d5
+	lsr.w	#1,d5	; get next bit
 	move	sr,d6
-	dbf	d4,+
+	dbf		d4,Kos_Decomp_ChkBit2
```
Comment (Clownacy) [inline s2.asm]: Same here.

File: s2.asm [hunk 6]
```diff
@@ -2293,88 +2335,99 @@ EniDec_ChkGetNextByte:
...
-	move	d6,ccr
-	bcs.s	+++
-	lsr.w	#1,d5
-	dbf	d4,+
+
+Kos_Decomp_ChkBit2:
+	move	d6,ccr	; was the bit set?
+	bcs.s	Kos_Decomp_FullMatch	; if it was, branch
+	lsr.w	#1,d5	; bit which is shifted out goes into X flag
+	dbf		d4,+
```
Comment (Clownacy) [inline s2.asm]: And here.

File: s2.asm [hunk 7]
```diff
@@ -2293,88 +2335,99 @@ EniDec_ChkGetNextByte:
...
 	move.b	(a0)+,1(sp)
 	move.b	(a0)+,(sp)
 	move.w	(sp),d5
 	moveq	#$F,d4
 +
-	roxl.w	#1,d3
+	roxl.w	#1,d3	; get high repeat count bit (shift X flag in)
 	lsr.w	#1,d5
-	dbf	d4,+
+	dbf		d4,+
```
Comment (Clownacy) [inline s2.asm]: This too.

File: s2.asm [hunk 8]
```diff
@@ -2293,88 +2335,99 @@ EniDec_ChkGetNextByte:
...
+	move.b	d1,d3	; copy repeat count
+	addq.w	#1,d3	; and increment it
+
+Kos_Decomp_MatchLoop:
 	move.b	(a1,d2.w),d0
-	move.b	d0,(a1)+
-	dbf	d3,-
-	bra.s	--
+	move.b	d0,(a1)+	; copy appropriate byte
+	dbf		d3,Kos_Decomp_MatchLoop	; and repeat the copying
```
Comment (Clownacy) [inline s2.asm]: This.

File: s2.asm [hunk 9]
```diff
@@ -798,9 +798,9 @@ SS_PNTA_Transfer_Table:	offsetTable
 	startZ80
 
 	bsr.w	ProcessDPLC2
-	tst.w	(Demo_Time_left).w
-	beq.w	+	; rts
-	subq.w	#1,(Demo_Time_left).w
+	tst.w	(Demo_Time_left).w	; is there time left on the demo?
+	beq.w	+	; if not, return
+	subq.w	#1,(Demo_Time_left).w	; subtract 1 from time left in demo
```
Comment (kirjavascript) [inline s2.asm]: shouldn't comments that are basically;
`x = x + 1; // add 1 to x`
be omitted?

Comment (Clownacy) [reply s2.asm]: If my memory's right, the other instances of this code already had comments like these, so it's more of a consistency thing.

Granted, I would prefer this disassembly have little of its fat trimmed, but that is a lot of code to sift through. You also have the fact that coding novices read through this.

File: s2.asm (hunk 3 above)
Comment (GabrielRavier) [reply s2.asm]: I though it looked fine but does that mean that ; =============== S U B R O U T I N E ======================================= appears wrong with 8 tabs ?

Comment (GabrielRavier) [reply s2.asm]: Oh sorry I though window size not tabulation size sorry

**Comment by GabrielRavier** [comment]
I corrected all my errors you pointed out

**Comment by flamewing** [comment]
And again with a generic commit message that does not have any relation to what the commit does.

**Comment by GabrielRavier** [comment]
I didn't remember exactly what were the contents of the changes. Adding actual commit message...

**Comment by GabrielRavier** [comment]
Also here is the complete changelog : 
1. Made a few changes to the header comments
2. "4 bytes" to "a long"
3. Removed things such as "subtract 1 from time left in demo"
4. Reformulated a few comments
5. Commented the controller-reading
6. Removed a tab in an empty space
7. Commented the Nemesis decompression routine
8. Commented the Enigma decompression routine
9. Commented the Kosinski decompression routine
10. Changed strange commenting in Super Sonic PAL cycle

**Comment by GabrielRavier** [comment]
Fixed that commit Clownacy

**Comment by GabrielRavier** [comment]
Closing this because I have no intent to correct this ever and I also am preparing another pull request so...

**Pull request closed without merging**